components/locid_transform/src/expander.rs

```diff
@@ -477,83 +511,165 @@ impl LocaleExpander {
+            None
+        };
 
-        trial.script = max.script;
-        trial.region = None;
-        self.maximize(&mut trial);
-        if trial == max {
-            if langid.language != max.language
-                || langid.script != max.script
-                || langid.region.is_some()
-            {
-                if langid.language != max.language {
-                    langid.language = max.language
+        let languagemax_scriptmax = |langid: &mut LanguageIdentifier| -> Option<TransformResult> {
+            let mut trial = max.clone();
+            trial.script = max.script;
+            trial.region = None;
+            self.maximize(&mut trial);
+            if trial == max {
+                if langid.language != max.language
+                    || langid.script != max.script
+                    || langid.region.is_some()
+                {
+                    if langid.language != max.language {
+                        langid.language = max.language
+                    }
+                    if langid.script != max.script {
+                        langid.script = max.script;
+                    }
+                    if langid.region.is_some() {
+                        langid.region = None;
+                    }
+                    return Some(TransformResult::Modified);
+                } else {
+                    return Some(TransformResult::Unmodified);
                 }
-                if langid.script != max.script {
-                    langid.script = max.script;
+            }
+            None
+        };
+
+        'fail: {
+            let res = 'success: {
+                if let Some(result) = languagemax(langid) {
+                    break 'success result;
                 }
-                if langid.region.is_some() {
-                    langid.region = None;
+                match options.subtag_priority {
+                    LocaleExpanderSubtagPriority::Script => {
+                        if let Some(result) = languagemax_scriptmax(langid) {
+                            break 'success result;
+                        }
+                        if let Some(result) = languagemax_regionmax(langid) {
+                            break 'success result;
+                        }
+                    }
+                    LocaleExpanderSubtagPriority::Region => {
+                        if let Some(result) = languagemax_regionmax(langid) {
+                            break 'success result;
+                        }
+                        if let Some(result) = languagemax_scriptmax(langid) {
+                            break 'success result;
+                        }
+                    }
                 }
-                langid.variants = variants;
-                return TransformResult::Modified;
-            } else {
-                return TransformResult::Unmodified;
-            }
+                break 'fail;
+            };
+            langid.variants = variants;
```

Comment:
`langid.variants` is restored here instead of inside the closures since that involves moving `variants` inside each closure, which would require cloning `variants`.

Reply:
I don't think we actually need to do anything special with `variants`. The `maximize` function already doesn't do anything with them. #4754